fdbrpc/include/fdbrpc/FileTransfer.h

```diff
@@ -0,0 +1,263 @@
+	uint32_t crc = 0;
+	std::vector<char> buffer(8192);
+	while (input_file->read(buffer.data(), buffer.size()) || input_file->gcount() > 0) {
+		crc = crc32c_append(crc, reinterpret_cast<const uint8_t*>(buffer.data()), input_file->gcount());
+	}
+	return crc;
+}
+
+class FileTransferClient {
+public:
+	FileTransferClient(std::shared_ptr<Channel> channel) : stub_(FileTransferService::NewStub(channel)) {}
+
+	std::optional<GetFileInfoReply> GetFileInfo(const std::string& filename, bool get_crc_checksum = false) {
+		ClientContext context;
+		GetFileInfoRequest request;
+		request.set_file_name(filename);
+		request.set_get_size(true);
+		request.set_get_crc_checksum(get_crc_checksum);
+
+		GetFileInfoReply response;
+		Status status = stub_->GetFileInfo(&context, request, &response);
+		if (status.ok()) {
+			return { response };
+		} else {
+			return std::nullopt;
+		}
+	}
+
+	// Downloads a file from a remote server and saves it locally.
+	//
+	// This function communicates with a remote server to download a file specified by `filename`
+	// and saves it to the local file system with the name `output_filename`. Optionally, it can
+	// verify the integrity of the downloaded file using a CRC checksum.
+	//
+	// Params:
+	// - filename: The name of the file to download from the remote server.
+	// - output_filename: The name of the file to save the downloaded content locally.
+	// - verify: Whether to verify the file's integrity using CRC checksum
+	//
+	// Returns:
+	//  std::optional<size_t> The size of the downloaded file in bytes if successful, or std::nullopt if the download
+	//  failed. `output_filename` is deleted on failure.
+	std::optional<size_t> DownloadFile(const std::string& filename,
+	                                   const std::string& output_filename,
+	                                   bool verify = true) {
+
+		uint32_t expected_crc = 0;
+		uint32_t expected_size = 0;
+		{
+			GetFileInfoRequest request;
+			ClientContext context;
+			request.set_file_name(filename);
+			request.set_get_crc_checksum(verify);
+			request.set_get_size(true);
+			GetFileInfoReply response;
+			auto res = stub_->GetFileInfo(&context, request, &response);
+			if (!res.ok()) {
+				return std::nullopt;
+			}
+			expected_crc = response.crc_checksum();
+			expected_size = response.file_size();
+		}
+
+		DownloadRequest request;
+		request.set_file_name(filename);
+
+		ClientContext context;
+		std::unique_ptr<grpc::ClientReader<DownloadChunk>> reader(stub_->DownloadFile(&context, request));
+
+		std::ofstream output_file(output_filename, std::ios::binary | std::ios::trunc);
+		if (!output_file.is_open()) {
+			return std::nullopt; // Failed to open file
+		}
+
+		DownloadChunk chunk;
+		size_t bytes_read = 0;
+		bool failed = false;
+		while (reader->Read(&chunk)) {
+			if (chunk.offset() != bytes_read) {
+				// Abort on invalid offset
+				failed = true;
+				break;
+			}
+
+			output_file.write(chunk.data().data(), chunk.data().size());
```

Comment:
This is not async. You thinking grpc on a background or dedicated thread?

Reply:
Yes. That's the plan. There is facility to call individual RPC async, but since we wanted entire thing to happen in background doing that didn't make sense, so it will go to a thread pool, which I will add later.

Reply:
Maybe add a comment somewhere to this effect... That this is your assumption?